f3.go

```diff
@@ -145,412 +78,197 @@ func New(ctx context.Context, manifest manifest.ManifestProvider, ds datastore.D
-	if err != nil {
-		return xerrors.Errorf("registering topic validator: %w", err)
+	if runner == nil {
+		m.log.Error("attempted to broadcast message while F3 wasn't running")
+		return
 	}
 
-	topic, err := m.pubsub.Join(pubsubTopicName)
+	err := runner.BroadcastMessage(msg)
 	if err != nil {
-		return xerrors.Errorf("could not join on pubsub topic: %s: %w", pubsubTopicName, err)
+		m.log.Errorf("failed to broadcast message: %+v", err)
 	}
-	m.client.topic = topic
-	return nil
 }
 
-func (m *F3) teardownPubsub(manifest manifest.Manifest) (err error) {
-	m.psLk.Lock()
-	defer m.psLk.Unlock()
-	if m.client.topic != nil {
-		m.msgSub.Cancel()
-		err = multierr.Combine(
-			m.pubsub.UnregisterTopicValidator(manifest.PubSubTopic()),
-			m.client.topic.Close(),
-		)
-		m.client.topic = nil
-	}
-	return err
-}
-
-func (m *F3) initGpbftRunner(ctx context.Context) error {
-	m.mLk.Lock()
-	defer m.mLk.Unlock()
-
-	cs, err := certstore.OpenStore(ctx, m.ds)
-	if err == nil {
-		m.setCertStore(cs)
-	} else if errors.Is(err, certstore.ErrNotInitialized) {
-		err := m.boostrap(ctx)
-		if err != nil {
-			return xerrors.Errorf("failed to boostrap: %w", err)
-		}
-	} else {
-		return xerrors.Errorf("opening certstore: %w", err)
-	}
-
-	m.runner, err = newRunner(m.client.manifest, m.client)
-	if err != nil {
-		return xerrors.Errorf("creating gpbft host: %w", err)
-	}
-
-	m.psLk.Lock()
-	if err := m.setupPubsub(m.runner); err != nil {
-		return xerrors.Errorf("setting up pubsub: %w", err)
-	}
-	m.msgSub, err = m.client.topic.Subscribe()
-	if err != nil {
-		m.psLk.Unlock()
-		return xerrors.Errorf("subscribing to topic: %w", err)
-	}
-	m.psLk.Unlock()
-	return nil
-}
+func (m *F3) GetLatestCert(ctx context.Context) (*certs.FinalityCertificate, error) {
+	m.mu.Lock()
+	cs := m.cs
+	m.mu.Unlock()
 
-// Sets up the gpbft runner, this is triggered at initialization
-func (m *F3) startGpbftRunner(ctx context.Context, errCh chan error) {
-	if err := m.initGpbftRunner(ctx); err != nil {
-		errCh <- xerrors.Errorf("initializing gpbft host: %w", err)
-		return
+	if cs == nil {
+		return nil, xerrors.Errorf("F3 is not running")
 	}
-
-	// the size of the buffer is set to prevent message spamming from pubsub,
-	// so it is a big enough buffer to be able to accommodate new messages
-	// at "high-rate", but small enough to avoid spamming or clogging the node.
-	messageQueue := make(chan gpbft.ValidatedMessage, 20)
-	m.client.messageQueue = messageQueue
-
-	go func() {
-		m.mLk.Lock()
-		latest := m.certStore.Latest()
-		m.mLk.Unlock()
-		startInstance := m.Manifest().InitialInstance
-		if latest != nil {
-			startInstance = latest.GPBFTInstance + 1
-		}
-		// Check context before starting the instance
-		select {
-		case <-ctx.Done():
-			// If the context is cancelled before starting, return immediately
-			errCh <- ctx.Err()
-			return
-		default:
-			// Proceed with running the instance
-			err := m.runner.Run(startInstance, ctx)
```

Comment:
This was racy. Given a quick re-config (e.g., catching up with reconfigs), we could start with the wrong instance if two copies of this function ended up running at the same time.